File: zemiao/209.py
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        start, end, n = 0,0,len(nums)
        res = 0
        if not nums:
          return 0
        temp = 0
        while end < n:
          temp += nums[end]
          while temp >= s:
            res = min(end-start+1,res) if res > 0 else end-start+1
            temp -= nums[start]
            start += 1
          end += 1
        return res

File: tnzhou/209.cpp
class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int leftlimit = 0;
        int curTotal = 0;
        int minLength = nums.size() + 1;
        
        for (int i = 0; i < nums.size(); i ++) {
            curTotal += nums[i];
            while (curTotal - nums[leftlimit] >= s) {
                curTotal -= nums[leftlimit];
                leftlimit ++;
            }
            if (curTotal >= s && i - leftlimit + 1 < minLength) {
                minLength = i - leftlimit + 1;
            }
        }
        
        return minLength > nums.size() ? 0 : minLength;
    }
};
